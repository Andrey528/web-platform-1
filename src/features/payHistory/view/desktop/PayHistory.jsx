import React from 'react';
import block from 'bem-cn';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import Spinner from 'components/Spinner';
import { actions as payHistoryActions } from '../../redux';
import './PayHistory.scss';

export class PayHistory extends React.Component {
  static propTypes = {
    locale: PropTypes.object.isRequired,
    history: PropTypes.array.isRequired,
    currency: PropTypes.string.isRequired,
    actionProcessing: PropTypes.bool.isRequired,
    
    getPayHistory: PropTypes.func.isRequired,
  }

  componentDidMount() {
    this.props.getPayHistory();
  }

  render() {
    const b = block('pay-history');
    const { locale, history, currency, actionProcessing } = this.props;
    const items = history.map(item => <div className={b('table-row')} key={item.id}>
      <span className={b('table-cell', { type: 'date' })}>{item.date}</span>
      <span className={b('table-cell', { type: 'type' })}>
        {item.type === 'payment' ? locale.paymentTypes[item.status] : `${item.paymentMode}: ${item.purse}`}
      </span>
      <span className={b('table-cell', { type: 'amount' })}>{`${item.amount} ${currency}`}</span>
      <span className={b('table-cell', { type: 'status' })}>
        {item.type === 'payment' ? locale.success : locale.withdrawalStatuses[item.status]}
      </span>
    </div>);
    return (
      <section className={b()}>
        <Spinner isLoading={actionProcessing} />
        <h4 className={b('title')}>{locale.payHistory}</h4>
        <div className={b('table')}>
          <header className={b('table-header')}>
            <span className={b('table-cell', { type: 'date' })}>{locale.date}</span>
            <span className={b('table-cell', { type: 'type' })}>{locale.operationType}</span>
            <span className={b('table-cell', { type: 'amount' })}>{`${locale.amount} ${currency}`}</span>
            <span className={b('table-cell', { type: 'status' })}>{locale.status}</span>
          </header>
          <main className={b('table-main')}>{items}</main>
        </div>
      </section>
    );
  }
}

function mapStateToProps(state) {
  return {
    locale: state.locale.payment,
    history: state.payHistory.history,
    currency: state.auth.user.currency,
    actionProcessing: state.payHistory.actionProcessing,
  };
}

function mapDispatchToProps(dispatch) {
  const actions = {
    getPayHistory: payHistoryActions.getPayHistory,
  };
  return bindActionCreators(actions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PayHistory);
