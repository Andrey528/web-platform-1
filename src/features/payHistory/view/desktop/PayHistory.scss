@import 'shared/style/varDesktop.scss';

.pay-history {
  color: #000000;
  padding: 0.75rem 1.125rem;
  box-sizing: border-box;
  width: 100%;
  position: relative;
  height: 100%;
  font-size: 0.875rem;
  &__title {
    font-size: 1.5rem;
    font-weight: 400;
    margin-bottom: 0.8125rem;
  }
  &__table {
    width: 100%;
  }
  &__table-main {
    width: 100%;
  }
  &__table-row {
    display: flex;
    width: 100%;
    background: #ffffff;
    border-radius: 0.625rem;
    height: 2.0625rem;
    align-items: center;
    padding: 0 0.9375rem;
    box-sizing: border-box;
    margin: 0.3125rem 0;
    border: 1px solid $grey-border;
  }
  &__table-header {
    display: flex;
    width: 100%;
    background: #c12b2c;
    border-radius: 0.625rem;
    height: 2.0625rem;
    align-items: center;
    padding: 0 0.9375rem;
    color: #ffffff;
    box-sizing: border-box;
  }
  &__table-cell {
    height: 100%;
    display: flex;
    align-items: center;
    white-space: nowrap;
    &_type {
      &_date, &_amount, &_status {
        width: 15%;
      }
      &_type {
        width: 55%;
      }
      &_status {
        justify-content: center;
      }
    }
  }
}