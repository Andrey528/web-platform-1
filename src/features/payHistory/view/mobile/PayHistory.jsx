import React, { useEffect, useMemo } from 'react';
import block from 'bem-cn';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import BackBlock from 'components/BackBlock/mobile';
import Spinner from 'components/Spinner';
import { PayHistoryItem } from './PayHistoryItem/PayHistoryItem';
import { actions as payHistoryActions } from '../../redux';
import './PayHistory.scss';

export const PayHistory = ({ locale, localeCommon, history, actionProcessing, getPayHistory }) => {
  const b = block('pay-history');

  useEffect(() => {
    getPayHistory();
  }, [getPayHistory]);

  const items = useMemo(() => history.map(item =>
    <PayHistoryItem key={item.id} item={item} locale={locale} />), [history, locale]);
  return (
    <React.Fragment>
      <BackBlock>
        <span>{localeCommon.payHistory}</span>
      </BackBlock>
      <section className={b()}>
        <Spinner isLoading={actionProcessing} />
        {items}
      </section>
    </React.Fragment>
  );
};

PayHistory.propTypes = {
  locale: PropTypes.object.isRequired,
  localeCommon: PropTypes.object.isRequired,
  history: PropTypes.array.isRequired,
  actionProcessing: PropTypes.bool.isRequired,

  getPayHistory: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    locale: state.locale.payment,
    localeCommon: state.locale.common,
    history: state.payHistory.history,
    actionProcessing: state.payHistory.actionProcessing,
  };
}

function mapDispatchToProps(dispatch) {
  const actions = {
    getPayHistory: payHistoryActions.getPayHistory,
  };
  return bindActionCreators(actions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PayHistory);
