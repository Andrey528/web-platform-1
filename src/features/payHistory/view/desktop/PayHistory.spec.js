import React from 'react';
import { shallow, mount } from 'enzyme';
import block from 'bem-cn';

import { PayHistory } from './PayHistory';

describe('PayHistory tests', () => {
  const b = block('pay-props.history');
  const props = {
    locale: {
      success: 'success test',
      payHistory: 'payHistory test',
      date: 'date test',
      operationType: 'operationType test',
      amount: 'amount test',
      status: 'status test',
      withdrawalStatuses: { 0: 'Created', 1: 'During', 2: 'Paid out', 3: 'Blocked', 4: 'Canceled' },
      paymentTypes: { 1: 'Refill balance', 6: 'Bet win', 10: 'Withdrawal', 16: 'Payment with FREEKASSA', 114: 'Payment with cashier' },
    },
    history: [
      { date: '05.08.2019 09:15', id: 1751, status: 6, amount: 60, type: 'payment' },
      { date: '24.07.2019 02:14', id: 480, status: 6, amount: 25, type: 'payment' },
      { date: '18.07.2019 21:43', id: 10, paymentMode: 'Card', purse: null, status: 1, amount: 50, type: 'withdrawal' },
      { date: '11.07.2019 18:33', id: 7, paymentMode: 'Card', purse: '50505050', status: 1, amount: 50, type: 'withdrawal' },
    ],
    currency: 'RUB',
    actionProcessing: false,
    getPayHistory: jest.fn(),
  };

  describe('PayHistory with props', () => {
    const payHistory = shallow(<PayHistory {...props} />);
    const { locale, currency } = props;

    it('render properly', () => {
      expect(payHistory).toMatchSnapshot();
    });

    it('render title', () => {
      expect(payHistory.find('.pay-history__title').text()).toEqual(locale.payHistory);
    });

    describe('render table header text', () => {
      it('render table header date', () => {
        expect(payHistory.find('.pay-history__table-cell_type_date').at(0).text())
          .toEqual(locale.date);
      });

      it('render table header type', () => {
        expect(payHistory.find('.pay-history__table-cell_type_type').at(0).text())
          .toEqual(locale.operationType);
      });

      it('render table header amount', () => {
        expect(payHistory.find('.pay-history__table-cell_type_amount').at(0).text())
          .toEqual(`${locale.amount} ${currency}`);
      });

      it('render table header status', () => {
        expect(payHistory.find('.pay-history__table-cell_type_status').at(0).text())
          .toEqual(locale.status);
      });
    });

    describe('render table cells text', () => {
      ['date', 'amount'].forEach(type => {
        it(`render table cell ${type}`, () => {
          props.history.forEach((key, i) => {
            expect(payHistory.find(`.pay-history__table-cell_type_${type}`).at(i + 1).text())
              .toEqual(type === 'amount' ? `${props.history[i][type]} ${currency}` : props.history[i][type]);
          });
        });
      });

      it('render table cell type', () => {
        props.history.forEach((key, i) => {
          expect(payHistory.find('.pay-history__table-cell_type_type').at(i + 1).text())
            .toEqual(key.type === 'payment' ? locale.paymentTypes[key.status] : `${key.paymentMode}: ${key.purse}`);
        });
      });

      it('render table cell status', () => {
        props.history.forEach((key, i) => {
          expect(payHistory.find('.pay-history__table-cell_type_status').at(i + 1).text())
            .toEqual(key.type === 'payment' ? locale.success : locale.withdrawalStatuses[key.status]);
        });
      });
    });

    describe('table rows count', () => {
      ['date', 'amount', 'type', 'status'].forEach(type => {
        it(`render table cells ${type} count`, () => {
          expect(payHistory.find(`.pay-history__table-cell_type_${type}`)).toHaveLength(props.history.length + 1);
        });
      });
    });

    it('render Spinner', () => {
      const nextProps = { ...props, actionProcessing: true };
      const component = shallow(<PayHistory {...nextProps} />);

      expect(component.find('Spinner')).toHaveLength(1);
    });

    it('should call componentDidMount once', () => {
      const componentDidMountSpy = jest.spyOn(PayHistory.prototype, 'componentDidMount');
      mount(<PayHistory {...props} />);
      expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
    });

    it('should call getPayHistory in componentDidMount', () => {
      mount(<PayHistory {...props} />);
      expect(props.getPayHistory).toHaveBeenCalledTimes(1);
    });
  });
});