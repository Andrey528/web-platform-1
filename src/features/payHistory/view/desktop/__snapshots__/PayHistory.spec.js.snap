// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayHistory tests PayHistory with props render properly 1`] = `
<section
  className="pay-history"
>
  <Spinner
    isLoading={false}
  />
  <h4
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    payHistory test
  </h4>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <header
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <span
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        date test
      </span>
      <span
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        operationType test
      </span>
      <span
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        amount test RUB
      </span>
      <span
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
      >
        status test
      </span>
    </header>
    <main
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="1751"
      >
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          05.08.2019 09:15
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          Bet win
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          60 RUB
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          success test
        </span>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="480"
      >
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          24.07.2019 02:14
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          Bet win
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          25 RUB
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          success test
        </span>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="10"
      >
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          18.07.2019 21:43
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          Card: null
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          50 RUB
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          During
        </span>
      </div>
      <div
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        key="7"
      >
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          11.07.2019 18:33
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          Card: 50505050
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          50 RUB
        </span>
        <span
          className={
            Object {
              "has": [Function],
              "is": [Function],
              "mix": [Function],
              "split": [Function],
              "state": [Function],
              "toString": [Function],
            }
          }
        >
          During
        </span>
      </div>
    </main>
  </div>
</section>
`;
